HIP/hip-358.md

```diff
@@ -0,0 +1,119 @@
+to rely solely on contract code for token management. 
+
+## User stories
+
+As a smart contract developer, I want to be able to create Hedera tokens through 
+solidity contract calls.
+
+## Specification
+
+### EVM Precompile extension
+
+The [Solidity file for development](../assets/hip-206/solidity/IHederaTokenService.sol) 
+is updated with new types and function signatures that the precompile will respond to. 
+It is included in this HIP by reference.
+
+### Solidity Function Signatures 
+
+Two functions will be added for creating fungible and non-fungible tokens each.
+One will be a simple version with just the required fields and the other a version 
+supporting custom fees as well. 
+The following structs will be added to simplify the interface between solidity 
+contract developers and the HTS precompile: 
+
+| Name                | Definition                                                                  |
+| ------------------- | ----------------------------------------------------------------------------|
+| `KeyValue`          | `(bool, address, bytes, bytes, address)`                                    |
+| `Expiry`            | `(uint, address, uint)`                                                     |
+| `TokenKey`          | `(uint, KeyValue)`                                                          |
+| `ThresholdTokenKey` | `(uint, uint, KeyValue[])`                                                  |
+| `HederaToken`       | `(string, string, address, string, bool, uint32, bool, TokenKey[], Expiry)` |
+| `FixedFee`          | `(uint32, address, address)`                                                |
+| `FractionalFee`     | `(uint32, uint32, uint32, uint32, bool, address)`                           |
+| `RoyaltyFee`        | `(uint32, uint32, FixedFee, address)`                                       |
+
+The ABI signature and hashes for each call are as follows:
+
+| hash       | signature                                                                                        | return                   |
+| ---------- | -------------------------------------------------------------------------------------------------|--------------------------|
+|  | `createFungibleToken(HederaToken, uint, uint)`                                                             | `(bool, addess, bytes)`  |
+|  | `createFungibleTokenWithCustomFees(HederaToken, uint, uint, FixedFee[], FractionalFee[])`                  | `(bool, addess, bytes)`  |
+|  | `createNonFungibleToken(HederaToken, uint, uint)`                                                          | `(bool, addess, bytes)`  |
+|  | `createNonFungibleTokenWithCustomFees(HederaToken, uint, uint, FixedFee[], FractionalFee[], RoyaltyFee[])` | `(bool, addess, bytes)`  |
+
+### Precompile Gas Costs
+
+Gas will be priced so that it will be cheaper to do calls directly to HTS rather
+than through HSCS where such calls are possible.
```

Comment:
We will need actual numbers before this can go final, but it won't block initial submission.

Can you add a table with the functiongroups and gas cost column with `TBD` as the price?